### DISCUSSION TITLE
Unable to get encryption working

### DISCUSSION THREAD

**Discussion by vaetas** [simolus3/drift#1026] [Q&A]
Hello,
I am trying to enable encryption according to the [docs](https://moor.simonbinder.eu/docs/other-engines/encryption/). Below is my code.

```dart
@UseMoor(tables: [])
class MoorDatabase extends _$MoorDatabase {
  MoorDatabase({
    @required String password,
  })  : assert(password != null),
        super(_openConnection(password));

  @override
  int get schemaVersion => 1;

  static LazyDatabase _openConnection(String password) {
    return LazyDatabase(() async {
      final databaseDirectory = await getApplicationDocumentsDirectory();
      final path = p.join(databaseDirectory.path, 'db.sqlite');

      return EncryptedExecutor(
        path: path,
        password: 'pass',
        logStatements: !kReleaseMode,
      );
    });
  }
}
```

However, I can change the `password` and the database will still be opened. I suppose that I'm missing some important step.

I tried to use the `moor/ffi` alternative. As I try this now, there seems to be the same problem.

```dart
return VmDatabase(
  File(path),
  setup: (rawDb) {
    rawDb.execute("PRAGMA key = 'pass'");
  },

  logStatements: !kReleaseMode,
);
```

Could you please point me in the right direction? Thanks!

Unlike the rest of the documentation, encryption chapter is not straightforward to follow. I think there should be more examples.

**Comment by davidmartos96** [comment]
@vaetas Have you actually inserted anything into the database? I believe SqlCipher only does the encryption after modifying the database.

**Reply by vaetas** [reply]
Yes, I tried inserting some data into the database.

But I think I found the problem. When I change the password inside the `EncryptedExecutor` and do a hot-restart, the database will still be unlocked (even with invalid password). However, when I change the password and completely stop & restart the app the database will not unlock with wrong password. I should have realized this sooner.

The `_openConnection(String password)` is still called on each hot-restart though. Should not a new connection be created?

**Reply by simolus3** [reply]
I think that might be related to #835? The app essentially gets restarted with a hot restart (`main` gets called again), so the Dart heap is lost and all objects are created again.

Since the process doesn't actually finish, this doesn't affect Flutter plugins or native code though. So it might be that an existing database is still managed by the `sqflite_sqlcipher` plugin and gets re-used. David probably knows more about this, but in general hot restarts can be problematic when managing native resources since there is no good way to dispose them.

**Reply by vaetas** [reply]
Yes, this is probably the reason. Thank you for the answer!
